api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -62,9 +62,38 @@ message DiscoveryRequest {
   // internal exception related to the failure. It is only intended for consumption during manual
   // debugging, the string provided is not guaranteed to be stable across Envoy versions.
   google.rpc.Status error_detail = 6;
+
+  // [#not-implemented-hide:]
+  // Latest xDS API patch version supported by the client.
+  // This is useful for use-cases where the client and server support new
+  // API features, before the next API minor version is rolled out.
+  // If not included, the patch version defaults to 0, stating that the client
+  // supports API features up to the client's minor version.
+  uint32 patch_version = 7;
```

Comment:
Why is this here instead of being in the `Node` message?  As currently designed, it's by definition something that works only when directly talking to an authoritative server, because it's not part of the resource name, so caching proxies can't handle it.

Note that if we do add a mechanism for flexible range-based context param matching, then we can make this part of the resource name, so it would work basically the same way as the minor version.  I think that would actually make this mechanism easier for people to understand.

Reply:
If it's not federated, it can live in the `Node` message agreed.

Reply:
I'm probably missing something but why isn't it effectively part of the resource name? Meaning, if we assume the patch level for a specific message is a recursive walk of all sub-messages, don't we know it on a per message/resource basis and then it might be different and should be set here? I guess maybe I'm then also confused why we don't need major/minor here also? Or we assume that those are sent in context params?

Reply:
I think it's related to the semantics of exact-match and of an optional patch support.
If this is part of the context-params, caches should only return resources with the patch, and caches will need to cache the same resource with different patch versions.

Reply:
OK thanks, this makes sense. Both this convo and the range convo does make me wonder a bit if we are thinking about context params wrong and the server should actually be able to return a result with a subset of matching params? Does that make sense?

Reply:
I think the core problem is that clients need a way to know which resources correspond to which of their requests.  If the server can send back resources that are just missing some of the params that the client ask for, then there's no way to unambiguously map the resources to the requests.

For example, let's say that a client has subscribed to the following:
- `xdstp:///envoy.config.route.v3.RouteConfiguration/route_config?my_context_param=foo`
- `xdstp:///envoy.config.route.v3.RouteConfiguration/route_config?my_context_param=bar`

If the server sends back a resource whose name is `xdstp:///envoy.config.route.v3.RouteConfiguration/route_config` with no `my_context_param` parameter, then how can the client know which of its requests that is the response to?

This is why I think that we're going to need clearly defined flexible matching semantics.  We need to be very sure that we don't accidentally introduce any ambiguity here.

Reply:
Yes, let's do range and move things to patch, thanks.

Reply:
What if there was a special wildcard param value? (Just thinking out loud here). This would then allow intermediate caches to be quite intelligent about mappings. Taking the patch value this would allow the server to respond with `patch_value=*` or whatever.

Reply:
I agree with moving patch into `Node`. Actually `Node` already has a field `ApiVersionNumber user_agent_latest_api_version` where a client can add the patch version it supports.
Just wanted to note that at the moment there's no support for patch version in the xdstp naming scheme version negotiation, and that should be fine because patch version support is optional.

> What if there was a special wildcard param value? (Just thinking out loud here). This would then allow intermediate caches to be quite intelligent about mappings. Taking the patch value this would allow the server to respond with `patch_value=*` or whatever.

IMO the _patch_ field should actually be understood as _maximal supported patch version_, so if the server responds with `patch_value=*`, the intermediate servers will not be able to know if the resource is supported in `max_patch_version=3` or in `max_patch_version=6`, and they'll need that to process future requests.

Reply:
> , the intermediate servers will not be able to know if the resource is supported in max_patch_version=3 or in max_patch_version=6, and they'll need that to process future requests.

But isn't that effectively by design? The management server shouldn't send * unless it really means it. I think what I'm getting at is it would give a way for the management server to effectively ignore patch that is cache compatible.

Reply:
I think it's possible to introduce a notion of cache-ignored context params. This is analogous to the ability in CDNs to ignore a list of query params. We've discussed adding this a few times to the spec.. I think the point becomes moot if we have range params though, since instead of max supported version N we can just say `[0..N]` with the same semantics and caches basically ignore it.